### DISCUSSION TITLE
How handle rm in pre-service

### DISCUSSION THREAD

**Discussion by realroot2185** [davmac314/dinit#409] [Q&A] [answered]
```
type = scripted
command = /usr/bin/rm -f $XDG_RUNTIME_DIR/bonsai
restart = false

type      = process
command   = /usr/bin/bonsaid -t $XDG_CONFIG_HOME/sxmo/bonsai_tree.json
waits-for = bonsaid-pre
logfile   = $HOME/.local/state/dinit/bonsai.log
restart   = true
```

Do I need to make a script that check if the file exists before deleting it?
To avoid pre-service failure.

Original service uses [superd](https://sr.ht/~craftyguy/superd/) that has "pre-command" and "on-failure" which can be used to send a notification or something for failed user services.

**Comment by mobin-2008** [comment]
Hello.

From [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/latest/#variables):
> $XDG_RUNTIME_DIR defines the base directory relative to which user-specific non-essential runtime files and other file objects (such as sockets, named pipes, ...) should be stored. The directory MUST be owned by the user, and they MUST be the only one having read and write access to it. Its Unix access mode MUST be 0700.

> The lifetime of the directory MUST be bound to the user being logged in. It MUST be created when the user first logs in and if the user fully logs out the directory MUST be removed. If the user logs in more than once they should get pointed to the same directory, and it is mandatory that the directory continues to exist from their first login to their last logout on the system, and not removed in between. Files in the directory MUST not survive reboot or a full logout/login cycle. 

~So there is no need for deleting it on every login.
However something like this could be used for ignoring rm failure:~
```
type = scripted
command = /bin/sh -c "/usr/bin/rm -f $XDG_RUNTIME_DIR/bonsai | :"
restart = false
```

**Reply by realroot2185** [reply]
I experienced bonsai crashing and it was not able to start cause the file was there.
Because the user did not log out.

In `command` can you use shell-like operations etc.?

**Comment by davmac314** [comment]
```
Do I need to make a script that check if the file exists before deleting it?
To avoid pre-service failure.
```

"rm -f" doesn't fail if the file does not exist.

**Reply by realroot2185** [reply]
```
[     {X}] bonsaid (exit status: 255)
[     {-}] bonsaid-pre
```
I thought that `{-}` meant failed sorry.

I used `restart = false` because I saw it in dbus-pre service.
To have it executed it every time before bonsaid service can I simply set restart to true?

**Reply by mobin-2008** [reply, marked as answer]
You should create a script for bonsaid service which does the remove trick on every startup, Something like this:
```
#!/bin/sh
/usr/bin/rm -f $XDG_RUNTIME_DIR/bonsai
exec /usr/bin/bonsaid -t $XDG_CONFIG_HOME/sxmo/bonsai_tree.json
```

**Reply by realroot2185** [reply]
No way to trigger the pre service restart therefore?

What the location of user `scritps` folder?
`/etc/dinit.d/user/scripts` and `~/.config/dinit.d/scripts`?

**Reply by notramo** [reply]
> What the location of user scritps folder?

It shouldn't really matter, you can put it anywhere on your system. The `scripts/` is just a convention. I recommend putting it into `~/.config/dinit.d/scripts`, so it can be version-controlled together with the config.

**Reply by realroot2185** [reply]
I was wondering if there was a standard for system-wide location.
Since the service it's in a package, I put it in `/etc/dinit.d/user/scripts`.

`rm` absolute path is different in some distros so I just put `rm`.
